Pin Chromium revision for binary installs

DanielRyanSmith opened the pull request:
Part of #14456

This change allows the pinning of a Chromium revision that has been tested against the WPT GitHub check suite.

Cloud Functions are run on WPT's GCP account that identifies the most recent Chromium revision that is mutually available to all major platforms (Linux, Mac, Windows). This revision is updated as an [external artifact](https://storage.googleapis.com/wpt-versions/pinned_chromium_revision_NEW), and a test run is initiated on [this draft PR](https://github.com/web-platform-tests/wpt/pull/34652). If all checks are successful, the revision is pinned as an an additional [external artifact](https://storage.googleapis.com/wpt-versions/pinned_chromium_revision). When a user installs Chromium via `./wpt install`, The revision installed will use this revision unless specified otherwise.

The `--revision` flag is now available for invocations of `./wpt install`. This will allow the user to specify which Chromium revision to install. This flag allows for `pinned` and `latest` arguments, which install the pinned revision or latest available revision respectively.

Chromium binaries are now installed noting the revision they are associated with. If a webdriver is installed to match a Chromium binary, or vice versa, this revision will be referenced and used to match.

This change does NOT switch any testing to use Chromium instead of Chrome. The code for installing Chromium has been added to `tools/ci/run_tc.py` but should not be part of the CI process.

DanielRyanSmith committed:
Pin Chromium revision for install

foolip commented:
@DanielRyanSmith what I suggested was based on my understanding of what https://github.com/puppeteer/puppeteer/blob/main/utils/check_availability.js does.

Running `node utils/check_availability.js 997451 996554` shows that 996554 has all the binaries, but also reveals how sparse the builds are. I had assumed that we build each revision and that binaries are only rarely missing due to a build error or infra problem, but that isn't at all the case. Rather, it looks like we have continuous builders that always build the latest revision when they become free, with no coordination between them.

To help judge if this is good enough, could you list the mutually available revisions in the past week? Does it happen every day, or less often?

foolip commented on .github/workflows/update_chromium_revision.yml:
```diff
@@ -0,0 +1,36 @@
+        cache: 'pip'
+    - run:  pip install requests
+    - name: execute update_chromium_revision.py
+      run: python -m tools.wpt.update_chromium_revision
```
This is unusual, at least in WPT's CI setup, why is the script invoked like this?

DanielRyanSmith replied:
It was a way to handle the relative imports in the file. I've refactored things so this is no longer needed.

foolip commented on .github/workflows/update_chromium_revision.yml:
```diff
@@ -0,0 +1,36 @@
+name: update_chromium_revision
+on:
+  # Trigger everyday at 10 UTC.
```
```suggestion
  # Trigger every day at 10 UTC, or manually.
```

foolip commented on .github/workflows/update_chromium_revision.yml:
```diff
@@ -0,0 +1,36 @@
+    steps:
+    - name: Checkout
+      uses: actions/checkout@v3
+    - name: setup python
```
```suggestion
    - name: Set up Python
```

foolip commented on .github/workflows/update_chromium_revision.yml:
```diff
@@ -0,0 +1,36 @@
+      with:
+        python-version: '3.10'
+        cache: 'pip'
+    - run:  pip install requests
```
```suggestion
    - name: Install dependencies
      run: pip install requests
```

foolip commented on .github/workflows/update_chromium_revision.yml:
```diff
@@ -0,0 +1,36 @@
+        token: ${{ secrets.GITHUB_TOKEN }}
+        author: wpt-pr-bot <wpt-pr-bot@users.noreply.github.com>
+        title: "Update latest available Chromium revisions"
+        commit-message: "Latest Chromium revision for Linux/Mac/Win"
```
```suggestion
        commit-message: "Latest Chromium revision for Linux/macOS/Windows"
```

foolip commented on tools/ci/run_tc.py:
```diff
@@ -257,13 +257,33 @@ def setup_environment(args):
         start_userspace_oom_killer()
 
 
+def install_chromium():
```
How about adding way to install the pinned revision using `./wpt install` and using that here?

If the pinned Chromium is only used (usable) by CI, then we should put `latest_chromium_revision.txt` somewhere under `tools/ci/`.

DanielRyanSmith replied:
I implemented it this way because we've always downloaded Chrome for testing in a way that is separate from `./wpt install`. If it is best for them to use the same install method, I can make that change.

foolip replied:
I think that's because Chrome can't be downloaded and installed easily on all platforms, and `./wpt install` should presumably work for all users. In CI we only need to support Linux.

Is the code path for installing Firefox also different? If not, the. Following that pattern should work. But if even Firefox isn't a single code path, then I'd not bother for Chromium either.

foolip commented on tools/wpt/browser.py:
```diff
@@ -529,13 +529,12 @@ class ChromeChromiumBase(Browser):
-        revision_url = ("https://storage.googleapis.com/chromium-browser-snapshots/"
-                        f"{architecture}/LAST_CHANGE")
-        return get(revision_url).text.strip()
+    def _get_latest_chromium_revision(self):
```
I would expect this to return an int, but it returns a string. Type annotation to make that very clear:

```suggestion
    def _get_pinned_chromium_revision(self) -> str:
```

You could also parse it to an int, but a string is OK.

Also suggest avoiding "latest", since it may no longer be.

foolip commented on tools/wpt/browser.py:
```diff
@@ -529,13 +529,12 @@ class ChromeChromiumBase(Browser):
-                        f"{architecture}/LAST_CHANGE")
-        return get(revision_url).text.strip()
+    def _get_latest_chromium_revision(self):
+        """Returns the latest Chromium revision number for the current platform."""
```
```suggestion
        """Returns the pinned Chromium revision number."""
```

foolip commented on tools/wpt/update_chromium_revision.py:
```diff
@@ -0,0 +1,88 @@
+
+PLATFORM_INFO = [
+    ("Win_x64", "chrome-win.zip"),
+    ("Win", "chrome-win.zip"),
```
These are the same 4 platforms that Puppeteer checks availability for, but if aligned builds are too rare, dropping win32 should make it about 10x as frequent.

Do you know when the 32-bit windows builds would end up being used? Does it require an old version of windows?

DanielRyanSmith replied:
It looks like some versions of Windows can come in 32-bits, [up to (rarely) Windows 10](https://www.engadget.com/windows-10-32-bit-oem-173055990.html). This seems really unusual though.

We do technically support installing for 32-bit Windows: 
https://github.com/web-platform-tests/wpt/blob/40e0b5cb8163efe413b448d7682629ac281b13da/tools/wpt/browser.py#L601-L611

This would remove this functionality. If that's okay - the win-32 lookup can definitely be removed.

foolip replied:
Let's drop it only if that helps get faster updates. It seems like maybe aligned binaries aren't as rare as we first feared.

DanielRyanSmith replied:
based on the week interval tested earlier (995555-998216)
```
Total viable revisions found: 33
Total viable revisions minus win32: 59
```

So not a major difference. Maybe worth keeping for now in this scenario?

foolip replied:
Yeah, let's keep it.

foolip commented on tools/wpt/update_chromium_revision.py:
```diff
@@ -0,0 +1,88 @@
+
+    # Step backwards through revision numbers until we find one
+    # that is available for all platforms.
+    largest_mutually_available_revision = smallest_revision
```
I think `largest_mutually_available_revision` is a long and slightly misleading name, since it doesn't actually become the largest mutually available revision until the very end. Suggesting a shorter name:

```suggestion
    candidate_revision = smallest_revision
```

foolip commented on tools/wpt/update_chromium_revision.py:
```diff
@@ -0,0 +1,88 @@
+    with open("tools/wpt/latest_chromium_revision.txt", "r") as f:
+        existing_revision = int(f.read())
+
+    # Find the lowest new revision number among latest revisions by platform.
```
This isn't wrong, but it's a bit of an unnecessary optimization. How about just starting with the Linux revision and iterating from there? More requests might be needed to find a match, but it's simpler.

DanielRyanSmith replied:
> This isn't wrong

Taking a look again at the logic, I actually _did_ implement it wrong, as it's taking the `max` of the revisions instead of the `min`. 😅 

I've changed this to be implemented as you said - starting with Linux and just going there for simplicity's sake.

foolip commented on tools/wpt/update_chromium_revision.py:
```diff
@@ -0,0 +1,88 @@
+                    available_for_all = False
+                    break
+            except requests.RequestException as e:
+                print(e)
```
Can you add a "bad thing happened when..." style description, that makes it clear whether it's safe to ignore this error? Or should we just let the script fail when there's a network error?

DanielRyanSmith replied:
Changed to better reflect what is happening 👍

foolip commented on tools/wpt/update_chromium_revision.py:
```diff
@@ -0,0 +1,88 @@
+                       f"{largest_mutually_available_revision}/{filename}")
+                # Check the headers of each possible download URL.
+                r = requests.head(url)
+                # If the "Accept-Ranges" header is not present, we know the file is not
```
Can we check the status code instead? Like this:

https://github.com/puppeteer/puppeteer/blob/94cb08c85955c0688d12b6ed10e61a4581a01280/src/node/BrowserFetcher.ts#L247-L271

DanielRyanSmith replied:
Yes - it should have been like this in the first place. 😅

foolip commented on tools/wpt/update_chromium_revision.py:
```diff
@@ -0,0 +1,88 @@
+
+    end = time()
+    if largest_mutually_available_revision == existing_revision:
+        print(f"No new mutually available revision numbers found after "
```
```suggestion
        print(f"No new mutually available revision found after "
```

DanielRyanSmith committed:
Apply suggestions from @foolip code review

DanielRyanSmith commented on tools/wpt/update_chromium_revision.py:
```diff
@@ -0,0 +1,88 @@
+                           f"{largest_mutually_available_revision}")
+
+    end = time()
+    if largest_mutually_available_revision == existing_revision:
```
Added

DanielRyanSmith commented:
@foolip
> To help judge if this is good enough, could you list the mutually available revisions in the past week? Does it happen every day, or less often?

I checked all revisions that were added in a roughly 1-week span from April 25th to March 2nd (995555-998216) and found 33 revisions that had a download available for every platform. That's a larger number than I expected, and perhaps the previous period we tested for had an unusually low number of viable revisions. If it is the case that we cannot find a new revision for a few days, will that be a significant problem?

Also, if it's of interest, these are the numbers that were found to be viable:
```
998197
998195
998186
998175
998162
998161
998157
998147
998145
998133
998132
998130
998124
998120
998097
998086
998046
998044
997617
997580
997535
996554
996439
996121
996032
995932
995889
995686
995685
995684
995683
995624
995589
```

ineo00048 commented:
It is done at 12:00 if not done early in no one can when requests are
available,user is the person able to make further decisions,also Infosys
involved 12:00 morning and mid morning before sunrise check for the tos
with the question?.

On Fri, Apr 29, 2022, 05:52 Philip Jägenstedt ***@***.***>
wrote:

> @DanielRyanSmith <https://github.com/DanielRyanSmith> what I suggested
> was based on my understanding of what
> https://github.com/puppeteer/puppeteer/blob/main/utils/check_availability.js
> does.
>
> Running node utils/check_availability.js 997451 996554 shows that 996554
> has all the binaries, but also reveals how sparse the builds are. I had
> assumed that we build each revision and that binaries are only rarely
> missing due to a build error or infra problem, but that isn't at all the
> case. Rather, it looks like we have continuous builders that always build
> the latest revision when they become free, with no coordination between
> them.
>
> To help judge if this is good enough, could you list the mutually
> available revisions in the past week? Does it happen every day, or less
> often?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/web-platform-tests/wpt/pull/33855#issuecomment-1113277087>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AYVS6KDHONWZCYPGEK2OA3TVHPLSRANCNFSM5UUC2YGQ>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

foolip commented on tools/ci/run_tc.py:
```diff
@@ -257,13 +257,33 @@ def setup_environment(args):
+    revisions_path = os.path.join(os.getcwd(), "tools", "wpt", "latest_chromium_revision.txt")
+    with open(revisions_path) as f:
+        revision = f.read().strip()
+        dest = os.path.join("/tmp")
```
Isn't this the same as just "/tmp"?

foolip commented on tools/ci/run_tc.py:
```diff
@@ -257,13 +257,33 @@ def setup_environment(args):
+               f"{revision}/chrome-linux.zip")
+        resp = get(url)
+        with open(installer_path, "wb") as f:
+            f.write(resp.content)
```
Does this end up reading the whole file into memory and then to disk? Can it be streamed to disk, or unzipped directly from the network? Given how big the archives are it seems like it could make a measurable difference. Worth timing this before doing too much work optimizing it.

DanielRyanSmith replied:
I'll take a look into it. Additionally, I think it might be worth it to remove this CI installation change entirely from this PR so that we can implement everything in a separate PR involved in changing to using Chromium for the CI.

jgraham commented:
Is the plan here to get daily PRs to update Chromium? Unless those are merged automatically it feels like it's going to be quite spammy.

foolip commented:
@jgraham Yes, we'll start with merging them manually and then look at making wpt-pr-bot approve them and merge if CI passes. Would it address the spamminess if we make wpt-pr-bot not assign any human reviewers for these PRs?

foolip commented on .github/workflows/update_chromium_revision.yml:
```diff
@@ -0,0 +1,37 @@
+      with:
+        token: ${{ secrets.GITHUB_TOKEN }}
+        author: wpt-pr-bot <wpt-pr-bot@users.noreply.github.com>
+        title: "Update latest available Chromium revisions"
```
Can you update the commit/PR wording to match recent changes? For this line, maybe:

```suggestion
        title: "Update pinned Chromium revision"
```

I think the commit message should also be the same as the PR title.

jgraham commented:
Maybe? I expect I'll see it a bit anyway, and it creates a bunch of extra churn for syncs (that we also get from frequently updated deps like hypothesis, but still).

In theory it could be avoidable if you could instead get the latest revision written to an external artifact at a knowable URL (I assume the GH API enables this kind of thing, but I haven't looked in detail). The disadvantage would be that there's a race where different CI systems could test the same revision with different versions of Chromium (that's solveable within a specific CI system by making the testrunning jobs depend on a revision-fetching job, like we do for Firefox in taskcuster). 

Assuming we go ahead with this approach, what's the tradeoff between the pr-bot merging the PR and the GH action itself merging the PR? It seems like the latter would be simpler, but maybe has worse security properties?

DanielRyanSmith commented:
@jgraham, @foolip and I were speaking about this earlier and explicitly brought up the idea of keeping this revision as an external artifact like you mentioned. The initial hesitance there is that we want to ensure that the revision chosen will pass the CI before being changed. I have not looked into the possibility of having the GH action merge the PR, but if we can ensure there will be no issues with the CI when this revision is changed, then that is a probable approach.

DanielRyanSmith committed:
Add revision flag for installation

DanielRyanSmith committed:
docstring description change

DanielRyanSmith committed:
Pinned revision is now external artifact

The pinned Chromium revision now lives in cloud storage.
Updates will be checked for it daily via a cloud function.

DanielRyanSmith committed:
Use pinned revision if a revision isn't detected

DanielRyanSmith committed:
make PR in correct repo

DanielRyanSmith committed:
mypy changes

DanielRyanSmith committed:
mypy changes

DanielRyanSmith committed:
remove GH workflow

DanielRyanSmith committed:
update logic for obtaining pinned revision

foolip reviewed (approved):
If we could have a `--revision=latest` that uses the current behavior, I think this is good to go.

DanielRyanSmith committed:
Add case for --revision=latest flag

DanielRyanSmith committed:
comment change
